Don't validate constants in const propagation

tmiasko opened the pull request:
Validation is neither necessary nor desirable.

The constant validation is already omitted at mir-opt-level >= 3, so there there are not changes in MIR test output (the propagation of invalid constants is covered by an existing test in tests/mir-opt/const_prop/invalid_constant.rs).

tmiasko committed:
Don't validate constants before propagation

Validation is neither necessary nor desirable.

The validation is already omitted at mir-opt-level >= 3, so there there
are not changes in MIR test output (the propagation of invalid constants
is covered by an existing test in tests/mir-opt/const_prop/invalid_constant.rs).

tmiasko committed:
Use trimmed paths in constantant validation errors

The constant validation errors are user facing and should always be
emitted to the user - use trimmed path when constructing them.

oli-obk commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit d2f3831869cf0cef3b9b1c64533e60b69c84e4e4 with merge 4ee8fa53366b8f492b736ddd0d3b28d579df07b4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"d2f3831869cf0cef3b9b1c64533e60b69c84e4e4","merge_sha":"4ee8fa53366b8f492b736ddd0d3b28d579df07b4"} -->

RalfJung commented on compiler/rustc_const_eval/src/interpret/validity.rs:
```diff
@@ -36,16 +36,14 @@ macro_rules! throw_validation_failure {
             msg.push_str(", but expected ");
             write!(&mut msg, $($expected_fmt)*).unwrap();
         )?
-        let path = rustc_middle::ty::print::with_no_trimmed_paths!({
```
Oh, I hadn't even realized that this was added in the validity checker. Happy to see it gone!

RalfJung commented:
I'm happy with the interpreter value validity checker change, but can't really comment on the ConstProp diff -- I have no idea why this check was added, and whether it was ever needed for correctness.

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4498853508/jobs/7918214706)
Build commit: 4ee8fa53366b8f492b736ddd0d3b28d579df07b4 (`4ee8fa53366b8f492b736ddd0d3b28d579df07b4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4498853508/jobs/7918214706"},"merge_sha":"4ee8fa53366b8f492b736ddd0d3b28d579df07b4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4498853508/jobs/7918214706)
Build commit: 4ee8fa53366b8f492b736ddd0d3b28d579df07b4 (`4ee8fa53366b8f492b736ddd0d3b28d579df07b4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4498853508/jobs/7918214706"},"merge_sha":"4ee8fa53366b8f492b736ddd0d3b28d579df07b4"} -->

saethlin commented:
I do not understand the justification for this change. It doesn't make sense to me to justify changing the default behavior of the compiler based on the fact that we test in a special configuration with different behavior.

tmiasko committed:
Remove workaround for CastKind::Transmute from const prop

Since constants are no longer validated before propagation the
workaround is obsolete. Remove it.

apiraino commented:
hello visiting this PR, checking progress.

@tmiasko can you share some thoughts about [this comment ](https://github.com/rust-lang/rust/pull/109521#issuecomment-1481183071)? thanks

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4814096190/jobs/8571402639) [(plain)](https://github.com/rust-lang/rust/commit/4adb8fbda002dc9b31b56b242094c16788586b3c/checks/13053676704/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
.............ii.iiii.........ii.......i.................................

failures:

---- [mir-opt] tests/mir-opt/const_prop/transmute.rs stdout ----
19 +         _2 = const {0x1 as &Never};      // scope 2 at $DIR/transmute.rs:+1:30: +1:48
20 +                                          // mir::Constant
21 +                                          // + span: no-location
- +                                          // + literal: Const { ty: &Never, val: Value(Scalar(0x0000000000000001)) }
+ +                                          // + literal: Const { ty: &Never, val: Value(Scalar(0x00000001)) }
23           StorageLive(_3);                 // scope 1 at $DIR/transmute.rs:+2:5: +2:16
24           unreachable;                     // scope 1 at $DIR/transmute.rs:+2:11: +2:13


thread '[mir-opt] tests/mir-opt/const_prop/transmute.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/tests/mir-opt/const_prop/transmute.unreachable_ref.ConstProp.diff', src/tools/compiletest/src/runtest.rs:3553:21


failures:
    [mir-opt] tests/mir-opt/const_prop/transmute.rs
```

</details>

tmiasko committed:
Use EMIT_MIR_FOR_EACH_BIT_WIDTH in transmute mir-opt test

wesleywiser commented:
I looked through the history and couldn't figure out why I added the validation either. I also checked Zulip to see if there was some discussion or PMs around that time but didn't find anything. 

@bors r+

bors commented:
:pushpin: Commit d1bd1be66834d166a9cf509250ffad1de8ff2709 has been approved by `wesleywiser`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=wesleywiser d1bd1be66834d166a9cf509250ffad1de8ff2709 -->
<!-- homu: {"type":"Approved","sha":"d1bd1be66834d166a9cf509250ffad1de8ff2709","approver":"wesleywiser","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit d1bd1be66834d166a9cf509250ffad1de8ff2709 with merge 5133e154590991203c84702ef59482f729b19bbf...
<!-- homu: {"type":"BuildStarted","head_sha":"d1bd1be66834d166a9cf509250ffad1de8ff2709","merge_sha":"5133e154590991203c84702ef59482f729b19bbf"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4857383712/jobs/8659653788)
Approved by: wesleywiser
Pushing 5133e154590991203c84702ef59482f729b19bbf to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"wesleywiser","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4857383712/jobs/8659653788"},"merge_sha":"5133e154590991203c84702ef59482f729b19bbf"} -->

rust-timer commented:
Finished benchmarking commit (5133e154590991203c84702ef59482f729b19bbf): [comparison URL](https://perf.rust-lang.org/compare.html?start=1cb63572d271855a5ccef79a3e10a8015d00f6ad&end=5133e154590991203c84702ef59482f729b19bbf&stat=instructions:u).

### Overall result: ❌ regressions - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=1cb63572d271855a5ccef79a3e10a8015d00f6ad&end=5133e154590991203c84702ef59482f729b19bbf&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 1.8% | [1.8%, 1.8%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1cb63572d271855a5ccef79a3e10a8015d00f6ad&end=5133e154590991203c84702ef59482f729b19bbf&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -4.1% | [-5.7%, -2.4%] | 2     |
| Improvements ✅ <br /> (secondary) | -3.6% | [-4.3%, -2.9%] | 2     |
| All ❌✅ (primary)                 | -4.1% | [-5.7%, -2.4%] | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1cb63572d271855a5ccef79a3e10a8015d00f6ad&end=5133e154590991203c84702ef59482f729b19bbf&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=1cb63572d271855a5ccef79a3e10a8015d00f6ad&end=5133e154590991203c84702ef59482f729b19bbf&stat=size:linked_artifact)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.2%  | [0.1%, 0.2%]   | 2     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.1% | [-0.2%, -0.1%] | 3     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.0% | [-0.2%, 0.2%]  | 5     |
</details>

**Bootstrap**: 655.55s -> 657.612s (0.31%)
